File: algorithms/easy/points_that_intersect_with_cars.ex
defmodule Solution do
  @spec number_of_points(nums :: [[integer]]) :: integer
  def number_of_points(nums) do
    mp =
      nums
      |> Enum.reduce(%{}, fn [start, end_], acc ->
        acc
        |> Map.update(start, 1, &(&1 + 1))
        |> Map.update(end_ + 1, -1, &(&1 - 1))
      end)

    1..101
    |> Enum.reduce({0, 0}, fn k, {ans, ct} ->
      ct = ct + Map.get(mp, k, 0)

      if ct > 0 do
        {ans + 1, ct}
      else
        {ans, ct}
      end
    end)
    |> elem(0)
  end
end

File: algorithms/easy/points_that_intersect_with_cars.rs
impl Solution {
    pub fn number_of_points(nums: Vec<Vec<i32>>) -> i32 {
        let mut arr = [0; 102];
        for n in nums {
            arr[n[0] as usize] += 1;
            arr[(n[1] + 1) as usize] -= 1;
        }
        let mut ct = 0;
        let mut ans = 0;
        for n in arr {
            ct += n;
            if (ct > 0) {
                ans += 1;
            }
        }

        ans
    }
}
